runtime/protocol/http-client-engines/http-client-engine-crt/common/src/aws/smithy/kotlin/runtime/http/engine/crt/CrtHttpEngine.kt

```diff
@@ -168,3 +144,32 @@ public class CrtHttpEngine(public val config: CrtHttpEngineConfig) : HttpClientE
         }
     }
 }
+
+internal suspend fun HttpStream.sendChunkedBody(body: HttpBody) {
+    when (body) {
+        is HttpBody.SourceContent -> {
+            val source = body.readFrom()
+            val bufferedSource = source.buffer()
+
+            while (!bufferedSource.exhausted()) {
+                bufferedSource.request(CHUNK_BUFFER_SIZE)
+                this.writeChunk(bufferedSource.buffer.readByteArray(), isFinalChunk = bufferedSource.exhausted())
+            }
+        }
+        is HttpBody.ChannelContent -> {
+            val chan = body.readFrom()
+            var buffer = SdkBuffer()
+            val nextBuffer = SdkBuffer()
+
+            while (!chan.isClosedForRead) {
+                chan.read(buffer, CHUNK_BUFFER_SIZE)
```

Comment:
**Question**: Is it possible on the initial iteration for `isClosedForRead` to return `false` but then the first `read` to return `-1` as a result of suspension?

Reply:
Yes its possible, although wouldn't that be an empty body? We wouldn't be able to handle that if I understood CRT requirements on sending empty chunks this morning.

Reply:
Yeah it is possible. It's not necessarily an empty body, it could be a nonempty body that got closed prematurely for some reason. CRT wouldn't be able to handle it either way though. Should I try to handle this case and throw an exception or just let it blow up? It seems like a very unlikely thing to happen.

Reply:
I advocate for additional handling of this case. _When_ we have to debug a failure here, it'll be nice to have a clear indication what happened rather than a cryptic CRT exception.

Reply:
Alright, I added some logic to handle this case where the initial chunk hasn't been sent but the channel is closed. Thanks!